File: solution_21.cpp
#include <iostream>
#include <math.h>
using namespace std;

int divisors_sum(int number) {
  int sum = 1;
  int lim = (int) sqrt(number);
  for (int x = 2; x <= lim; x++) {
    if (number % x == 0) {
      sum += x;
      sum += number / x;
    }
  }
  return sum;
}

bool amicable_pair(int num1, int num2) {
  return divisors_sum(num1) == num2 && divisors_sum(num2) == num1;
}

int solver(int from, int to) {
  int sum = 0;
  
  for (int x = from; x <= to; x++) {
    int y = divisors_sum(x);
    if (y > x) {
      if (amicable_pair(x, y)) {
        cout << "(" << x << ", " << y << ")" << endl;
        sum += (x + y);
      }
    }
  }
  
  return sum;
}

int main() {
  cout << solver(1, 100000) << endl;
}

File: solution_21.rb
# find the sum of all amidable numbers smaller than 10 000

def divisors_sum number
  sum = 1
  lim = Math::sqrt(number).to_i
  
  for x in 2..lim
    if (number % x == 0)
      sum += x 
      sum += number / x      
    end
  end
  
  return sum
end

def amicable_pair?(num1, num2)
  divisors_sum(num1) == num2 and divisors_sum(num2) == num1
end

def solver(from, to)
  sum = 0
  
  for x in from..to
    y = divisors_sum(x)
    if (y > x) 
      if amicable_pair?(x, y)
        sum += (x + y) 
        puts "(#{x}, #{y})"
      end
    end
  end
  
  return sum
end

puts solver(1, 100000)
